tests/varispeed_test.c

```diff
@@ -144,8 +152,101 @@ varispeed_test (int converter, double target_snr)
 		exit (1) ;
 		} ;
 
-	puts ("ok") ;
-
 	return ;
 } /* varispeed_test */
 
+static void
+varispeed_hammer_test (int converter)
+{	double ratios [] = { 1.0, 0.01, 20 } ;
+	int chan, r1, r2 ;
+
+	for (chan = 1 ; chan <= 9 ; chan ++)
+		for (r1 = 0 ; r1 < ARRAY_LEN (ratios) ; r1++)
+			for (r2 = 0 ; r1 != r2 && r2 < ARRAY_LEN (ratios) ; r2 ++)
+				set_ratio_test (converter, 1, ratios [r1], ratios [r2]) ;
+} /* varispeed_hammer_test */
+
+static void
+set_ratio_test (int converter, int channels, double initial_ratio, double second_ratio)
+{	static float input [BUFFER_LEN] ;
+	static float output [8 * BUFFER_LEN] ;
+	static float temp_buffer [BUFFER_LEN] ;
+	static char details [128] ;
+
+	const int num_frames = BUFFER_LEN / channels ;
+	const int chunk_size = 128 ;
+
+	SRC_STATE *src_state ;
+	SRC_DATA src_data ;
+
+	long input_len, output_len, current_in, current_out ;
+	int error, k, ch ;
+
+
+
+	snprintf (details, sizeof (details), "%d channels, ratio %g -> %g", channels, initial_ratio, second_ratio) ;
+
+	input_len = ARRAY_LEN (input) / channels ;
+	output_len = ARRAY_LEN (output) / channels ;
+
+	for (ch = 0 ; ch < channels ; ch++)
+	{	double freq = 0.01 * (ch + 1) ;
+		gen_windowed_sines (1, &freq, 1.0, temp_buffer + ch * num_frames, num_frames) ;
+		} ;
+
+	interleave_data (temp_buffer, input, num_frames, channels) ;
+
+	if ((src_state = src_new (converter, channels, &error)) == NULL)
+	{	printf ("\n\nLine %d : src_new () failed : %s\n\n", __LINE__, src_strerror (error)) ;
+		exit (1) ;
+		} ;
+
+	current_in = 0 ;
+	current_out = 0 ;
+
+	memset (&src_data, 0, sizeof (src_data)) ;
+	src_data.end_of_input = 0 ;
+	src_data.src_ratio = initial_ratio ;
+	src_data.data_in = input ;
+	src_data.data_out = output ;
+	src_data.input_frames = chunk_size ;
+	src_data.output_frames = ARRAY_LEN (output) / channels ;
+
+	/* Process one chunk at initial_ratio. */
+	if ((error = src_process (src_state, &src_data)))
+	{	printf ("\n\nLine %d : %s : %s\n\n", __LINE__, details, src_strerror (error)) ;
+		exit (1) ;
+		} ;
+
+	/* Now switch to second_ratio and process the remaining. */
+	src_data.src_ratio = second_ratio ;
+
+	while (1)
+	{	if ((error = src_process (src_state, &src_data)))
+		{	printf ("\n\nLine %d : %s : %s\n\n", __LINE__, details, src_strerror (error)) ;
+			exit (1) ;
+			} ;
+
+		if (src_data.end_of_input && src_data.output_frames_gen == 0)
+			break ;
+
+		current_in	+= src_data.input_frames_used ;
+		current_out += src_data.output_frames_gen ;
+
+		src_data.data_in	+= src_data.input_frames_used ;
+		src_data.data_out	+= src_data.output_frames_gen ;
+
+		src_data.input_frames	= input_len - current_in ;
+		src_data.output_frames	= output_len - current_out ;
+
+		src_data.end_of_input = (current_in >= input_len) ? 1 : 0 ;
+		} ;
+
+
+	for (k = 0 ; k < current_out * channels ; k ++)
```

Comment:
Missing `ASSERT(current_out > 0)` which can cause a test success with no check of the output (in the case the buffers are to small to produce any output)